src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -496,14 +513,111 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
+        // and rebuilt in the test framework.
+        //
+        // This is complicated by the fact that user defined functions cannot
+        // take/return AbstractFloats, and AbstractFloats cannot be stored in
+        // variables, so the code cannot just inject a simple utility function
+        // at the top of the shader, instead this snippet needs to be inlined
+        // everywhere the test needs to return an AbstractFloat.
+        //
+        // select is used below, since ifs are not available during constant
+        // eval. This has the side effect of short-circuiting doesn't occur, so
+        // both sides of the select have to evaluate and be valid.
+        //
+        // This snippet implements FTZ for subnormals to bypass the need for
+        // complex subnormal specific logic.
+        //
+        // Expressions resulting in subnormals can still be reasonably tested,
+        // since this snippet will return 0 with the correct sign, which is
+        // always in the acceptance interval for a subnormal result, since an
+        // implementation may FTZ.
+        //
+        // Document for the snippet is included here in this code block, since
+        // shader length affects compilation time  significantly on some
+        // backends.
+        //
+        // Snippet with documentation:
+        //   const kExponentBias = 1022;
+        //
+        //   // Detect if the value is zero or subnormal, so that FTZ behaviour
+        //   // can occur
+        //   const subnormal_or_zero : bool = (${expr} <= ${kValue.f64.subnormal.positive.max}) && (${expr} >= ${kValue.f64.subnormal.negative.min});
+        //
+        //   // MSB of the upper u32 is 1 if the value is negative, otherwise 0
+        //   // Extract the sign bit early, so that abs() can be used with
+        //   // frexp() so negative cases do not need to be handled
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expr} < 0);
+        //
+        //   // Use frexp() to obtain the exponent and fractional parts, and
+        //   // then perform FTZ if needed
+        //   const f = frexp(abs(${expr}));
+        //   const f_fract = select(f.fract, 0, subnormal_or_zero);
+        //   const f_exp = select(f.exp, -kExponentBias, subnormal_or_zero);
+        //
+        //   // Adjust for the exponent bias and shift for storing in bits
+        //   // [20..31] of the upper u32
+        //   const exponent_bits : u32 = (f_exp + kExponentBias) << 20;
+        //
+        //   // Extract the portion of the mantissa that appears in upper u32 as
+        //   // a float for later use
+        //   const high_mantissa = ldexp(f_fract, 21);
+        //
+        //   // Extract the portion of the mantissa that appears in upper u32 as
+        //   // as bits. This value is masked, because normals will explicitly
+        //   // have the implicit leading 1 that, which shouldn't be in the
```

Comment:
nit: grammar  "have the implicit leading 1 that should not be in the final result".

Reply:
Done